src/lib/stable/canister_methods/index.ts

```diff
@@ -171,50 +193,99 @@ function decoratorImplementation<This, Args extends unknown[], Return>(
 
-            verifyInitAndPostUpgradeHaveTheSameParams(
-                exportedCanisterClassInstance._azleInitAndPostUpgradeIdlTypes
-            );
+            if (postUpgradeDefined === true) {
+                verifyInitAndPostUpgradeHaveTheSameParams(
+                    paramIdlTypes ?? [],
+                    exportedCanisterClassInstance._azleInitAndPostUpgradeIdlTypes
+                );
+            } else {
+                exportedCanisterClassInstance._azleInitAndPostUpgradeIdlTypes =
+                    paramIdlTypes ?? [];
+            }
         }
 
         if (canisterMethodMode === 'postUpgrade') {
+            throwIfMethodAlreadyDefined(
+                'postUpgrade',
+                exportedCanisterClassInstance._azleDefinedSystemMethods
+                    .postUpgrade
+            );
+
+            exportedCanisterClassInstance._azleDefinedSystemMethods.postUpgrade =
+                true;
             exportedCanisterClassInstance._azleMethodMeta.post_upgrade = {
                 name,
                 index
             };
 
-            exportedCanisterClassInstance._azleInitAndPostUpgradeIdlTypes.push(
-                IDL.Func(paramIdlTypes ?? [], [], ['post_upgrade'])
-            );
+            const initDefined =
+                exportedCanisterClassInstance._azleDefinedSystemMethods.init;
 
-            verifyInitAndPostUpgradeHaveTheSameParams(
-                exportedCanisterClassInstance._azleInitAndPostUpgradeIdlTypes
-            );
+            if (initDefined === true) {
+                verifyInitAndPostUpgradeHaveTheSameParams(
+                    paramIdlTypes ?? [],
+                    exportedCanisterClassInstance._azleInitAndPostUpgradeIdlTypes
+                );
+            } else {
+                exportedCanisterClassInstance._azleInitAndPostUpgradeIdlTypes =
+                    paramIdlTypes ?? [];
+            }
         }
 
         if (canisterMethodMode === 'preUpgrade') {
+            throwIfMethodAlreadyDefined(
+                'preUpgrade',
+                exportedCanisterClassInstance._azleDefinedSystemMethods
+                    .preUpgrade
+            );
+
+            exportedCanisterClassInstance._azleDefinedSystemMethods.preUpgrade =
+                true;
             exportedCanisterClassInstance._azleMethodMeta.pre_upgrade = {
                 name,
                 index
             };
         }
 
         if (canisterMethodMode === 'heartbeat') {
+            throwIfMethodAlreadyDefined(
+                'heartbeat',
+                exportedCanisterClassInstance._azleDefinedSystemMethods
+                    .heartbeat
+            );
+
+            exportedCanisterClassInstance._azleDefinedSystemMethods.heartbeat =
+                true;
             exportedCanisterClassInstance._azleMethodMeta.heartbeat = {
                 name,
                 index
             };
         }
 
         if (canisterMethodMode === 'inspectMessage') {
```

Comment:
Move inspectMessage above heartbeat everywhere, I like this order: init, post upgrade, pre upgrade, inspect message, heartbeat and it's the order we've been using in the Call Context documentation

Reply:
Done